File: go/protocols/mpp/server/session_voucher.go
package server

// Voucher verifier for the MPP session server.
//
// Pure function: given a current channel snapshot and a signed voucher,
// decide whether to accept (and what the new watermark would be), reject,
// or treat as an idempotent replay. The caller persists any accepted delta
// through ChannelStore.UpdateChannel, re-checking inside the atomic mutator.
//
// The check sequence (order and operators) is pinned across the language
// SDKs and harness-tested:
// parse u64 -> finalized -> close pending -> idempotent replay (same
// cumulative AND same signature, signature re-verified) -> cumulative >
// watermark strictly -> cumulative <= deposit -> delta >= minVoucherDelta ->
// Ed25519 verify against the stored authorizedSigner -> expiry.
//
// Expiry mirrors the on-chain settle guard, which rejects only
// `expires_at != 0 && now >= expires_at`: an expiresAt of 0 NEVER expires and
// is always accepted. A non-zero expiresAt is additionally required to outlast
// the settlement window (the channel's forced-close grace period): the server
// settles vouchers asynchronously, so a voucher that expires before the close
// transaction can land would be rejected on-chain after the server has already
// served. Rejecting `expiresAt < now + settlementWindow` keeps the off-chain
// accept decision consistent with what the program will accept at settle time.

import (
	"crypto/ed25519"
	"fmt"
	"strconv"
	"time"

	solana "github.com/gagliardetto/solana-go"

	"github.com/solana-foundation/pay-kit/go/protocols/mpp/intents"
)

// VoucherVerifyStatus is the outcome class of a voucher verification.
type VoucherVerifyStatus string

const (
	// VoucherVerifyAccepted means the voucher advanced the channel watermark.
	VoucherVerifyAccepted VoucherVerifyStatus = "accepted"

	// VoucherVerifyReplayed means an already-accepted voucher was re-submitted
	// (idempotent).
	VoucherVerifyReplayed VoucherVerifyStatus = "replayed"

	// VoucherVerifyRejected means the voucher was rejected; see
	// VoucherRejectReason.
	VoucherVerifyRejected VoucherVerifyStatus = "rejected"
)

// VoucherRejectReason is a stable string tag for voucher rejections so the
// caller can map to HTTP statuses / log levels without parsing free text.
// The tag values are stable across the language SDKs.
type VoucherRejectReason string

const (
	// VoucherRejectBelowMinDelta: the delta is below the configured minimum.
	VoucherRejectBelowMinDelta VoucherRejectReason = "below-min-delta"

	// VoucherRejectChannelClosePending: a close was already requested.
	VoucherRejectChannelClosePending VoucherRejectReason = "channel-close-pending"

	// VoucherRejectChannelFinalized: the channel is already finalized.
	VoucherRejectChannelFinalized VoucherRejectReason = "channel-finalized"

	// VoucherRejectCumulativeNotMonotonic: the cumulative does not strictly
	// exceed the watermark.
	VoucherRejectCumulativeNotMonotonic VoucherRejectReason = "cumulative-not-monotonic"

	// VoucherRejectExceedsDeposit: the cumulative exceeds the deposit cap.
	VoucherRejectExceedsDeposit VoucherRejectReason = "exceeds-deposit"

	// VoucherRejectExpired: the voucher expiry is not in the future.
	VoucherRejectExpired VoucherRejectReason = "expired"

	// VoucherRejectExpiryTooSoon: the voucher expiry is in the future but does
	// not outlast the settlement window, so the async settle transaction could
	// be rejected on-chain after the server has already served.
	VoucherRejectExpiryTooSoon VoucherRejectReason = "expiry-too-soon"

	// VoucherRejectInvalidCumulative: the cumulative does not parse as a u64.
	VoucherRejectInvalidCumulative VoucherRejectReason = "invalid-cumulative"

	// VoucherRejectInvalidSignature: the Ed25519 signature check failed.
	VoucherRejectInvalidSignature VoucherRejectReason = "invalid-signature"
)

// VoucherVerifyResult is the verdict of VerifyVoucherForChannel.
//
// Status selects which fields are meaningful: NewCumulative for accepted and
// replayed; NewExpiresAt and NewSignature for accepted only; Reason and
// Detail for rejected only.
type VoucherVerifyResult struct {
	// Status is the outcome class.
	Status VoucherVerifyStatus

	// NewCumulative is the watermark to persist (accepted) or the existing
	// watermark (replayed).
	NewCumulative uint64

	// NewExpiresAt is the expiry of the now-highest voucher (accepted only).
	NewExpiresAt int64

	// NewSignature is the signature to persist as HighestVoucherSignature
	// (accepted only, base58).
	NewSignature string

	// Reason is the stable rejection tag (rejected only).
	Reason VoucherRejectReason

	// Detail is a human-readable rejection detail. Safe to log; not stable.
	Detail string
}

// VerifyVoucherArgs are the inputs to VerifyVoucherForChannel.
type VerifyVoucherArgs struct {
	// State is the channel snapshot, typically read just before calling.
	State ChannelState

	// Signed is the voucher being submitted.
	Signed intents.SignedVoucher

	// Deposit is the authoritative deposit cap. Passed in (rather than read
	// off State) because some callers carry an updated cap after a recent
	// top-up that has not yet been written back into the store.
	Deposit uint64

	// MinVoucherDelta is the optional minimum delta from the previous
	// cumulative. Zero disables the check.
	MinVoucherDelta uint64

	// SettlementWindowSeconds is the channel's forced-close grace period in
	// seconds. A non-zero voucher expiry must outlast this window from now
	// (expiresAt >= now + settlementWindow) so the async settle transaction
	// cannot be rejected on-chain after the server has served. Zero applies the
	// plain `expiresAt > now` check only. Never affects an expiresAt of 0,
	// which never expires.
	SettlementWindowSeconds int64

	// NowSeconds overrides the clock (Unix seconds) for deterministic tests.
	// Nil defaults to time.Now().
	NowSeconds *int64
}

// VerifyVoucherForChannel verifies a voucher against a channel snapshot.
//
// Returns a verdict; the caller is responsible for persisting any accepted
// delta via ChannelStore.UpdateChannel. The verifier is pure: no store,
// network, or clock side effects (the clock is injectable).
func VerifyVoucherForChannel(args VerifyVoucherArgs) VoucherVerifyResult {
	state := args.State
	signed := args.Signed

	// 1. Parse new cumulative from the payload.
	newCumulative, err := strconv.ParseUint(signed.Data.Cumulative, 10, 64)
	if err != nil {
		return voucherReject(VoucherRejectInvalidCumulative,
			fmt.Sprintf("invalid cumulative in voucher: %s", signed.Data.Cumulative))
	}

	// 2. Channel must not be finalized.
	if state.Finalized {
		return voucherReject(VoucherRejectChannelFinalized,
			fmt.Sprintf("channel %s is already finalized", state.ChannelID))
	}

	// 3. Channel must not be in close-pending.
	if state.CloseRequestedAt != nil {
		return voucherReject(VoucherRejectChannelClosePending,
			fmt.Sprintf("channel %s close is pending; no further vouchers accepted", state.ChannelID))
	}

	// 4. Idempotent replay: same cumulative AND same signature. The signature
	// is re-verified so a replay of a forged voucher cannot slip through.
	if newCumulative == state.Cumulative &&
		state.HighestVoucherSignature != nil &&
		*state.HighestVoucherSignature == signed.Signature {
		if err := verifyVoucherSignatureBytes(signed, state.AuthorizedSigner); err != nil {
			return voucherReject(VoucherRejectInvalidSignature, err.Error())
		}
		if rejection := voucherExpiryRejection(signed.Data.ExpiresAt, voucherNow(args.NowSeconds), args.SettlementWindowSeconds); rejection != nil {
			return *rejection
		}
		return VoucherVerifyResult{Status: VoucherVerifyReplayed, NewCumulative: newCumulative}
	}

	// 5. Must strictly exceed the watermark (non-replay case).
	if newCumulative <= state.Cumulative {
		return voucherReject(VoucherRejectCumulativeNotMonotonic,
			fmt.Sprintf("voucher cumulative %d must exceed watermark %d", newCumulative, state.Cumulative))
	}

	// 6. Must not exceed the deposit.
	if newCumulative > args.Deposit {
		return voucherReject(VoucherRejectExceedsDeposit,
			fmt.Sprintf("voucher cumulative %d exceeds deposit %d", newCumulative, args.Deposit))
	}

	// 7. Min delta check.
	delta := newCumulative - state.Cumulative
	if args.MinVoucherDelta > 0 && delta < args.MinVoucherDelta {
		return voucherReject(VoucherRejectBelowMinDelta,
			fmt.Sprintf("voucher delta %d is below minimum %d", delta, args.MinVoucherDelta))
	}

	// 8. Verify the Ed25519 signature over the 48-byte canonical payload.
	if err := verifyVoucherSignatureBytes(signed, state.AuthorizedSigner); err != nil {
		return voucherReject(VoucherRejectInvalidSignature, err.Error())
	}

	// 9. Expiry. The caller may override NowSeconds for deterministic tests.
	if rejection := voucherExpiryRejection(signed.Data.ExpiresAt, voucherNow(args.NowSeconds), args.SettlementWindowSeconds); rejection != nil {
		return *rejection
	}

	return VoucherVerifyResult{
		Status:        VoucherVerifyAccepted,
		NewCumulative: newCumulative,
		NewExpiresAt:  signed.Data.ExpiresAt,
		NewSignature:  signed.Signature,
	}
}

// voucherReject builds a rejected verdict.
func voucherReject(reason VoucherRejectReason, detail string) VoucherVerifyResult {
	return VoucherVerifyResult{Status: VoucherVerifyRejected, Reason: reason, Detail: detail}
}

// voucherExpiryRejection applies the voucher expiry policy and returns a
// rejected verdict to surface, or nil when the expiry is acceptable.
//
// It mirrors the on-chain settle guard (`expires_at != 0 && now >=
// expires_at`): an expiresAt of 0 NEVER expires and is always accepted. A
// non-zero expiresAt must be strictly in the future, and when a settlement
// window is configured it must outlast that window from now so the async
// settle transaction cannot be rejected on-chain after the server has served.
func voucherExpiryRejection(expiresAt, now, settlementWindowSeconds int64) *VoucherVerifyResult {
	if expiresAt == 0 {
		// Never-expires voucher: matches the on-chain `expires_at == 0` case.
		return nil
	}
	if expiresAt <= now {
		rejection := voucherReject(VoucherRejectExpired, "voucher has expired")
		return &rejection
	}
	if settlementWindowSeconds > 0 && expiresAt < now+settlementWindowSeconds {
		rejection := voucherReject(VoucherRejectExpiryTooSoon,
			fmt.Sprintf("voucher expiry %d does not outlast the settlement window (now %d + %d)",
				expiresAt, now, settlementWindowSeconds))
		return &rejection
	}
	return nil
}

// voucherNow returns the override when set, otherwise the wall clock in Unix
// seconds.
func voucherNow(override *int64) int64 {
	if override != nil {
		return *override
	}
	return time.Now().Unix()
}

// verifyVoucherSignatureBytes checks the voucher's Ed25519 signature over the
// canonical 48-byte voucher payload against the authorized signer (both
// base58). The expiry check is not included; callers order it explicitly.
func verifyVoucherSignatureBytes(signed intents.SignedVoucher, authorizedSigner string) error {
	message, err := signed.Data.MessageBytes()
	if err != nil {
		return err
	}
	signature, err := solana.SignatureFromBase58(signed.Signature)
	if err != nil {
		return fmt.Errorf("invalid signature encoding: %w", err)
	}
	pubkey, err := solana.PublicKeyFromBase58(authorizedSigner)
	if err != nil {
		return fmt.Errorf("invalid authorized signer: %w", err)
	}
	if !ed25519.Verify(ed25519.PublicKey(pubkey.Bytes()), message, signature[:]) {
		return fmt.Errorf("voucher signature verification failed")
	}
	return nil
}

// verifySessionVoucher checks expiry (including the settlement-window margin)
// first, then the Ed25519 signature. Used by the commit and close paths; the
// voucher handler orders the two checks itself.
//
// An expiresAt of 0 NEVER expires (matching the on-chain settle guard and
// VerifyVoucherForChannel). A non-zero expiresAt must be strictly in the future
// AND, when settlementWindowSeconds > 0, outlast that window (now + window) so a
// committed or closing voucher can't expire before the async settle lands
// on-chain — the same guard VerifyVoucherForChannel applies on the voucher path.
func verifySessionVoucher(signed intents.SignedVoucher, authorizedSigner string, settlementWindowSeconds int64) error {
	if rejection := voucherExpiryRejection(signed.Data.ExpiresAt, time.Now().Unix(), settlementWindowSeconds); rejection != nil {
		return fmt.Errorf("%s", rejection.Detail)
	}
	return verifyVoucherSignatureBytes(signed, authorizedSigner)
}

File: python/src/solana_pay_kit/protocols/mpp/server/session_voucher.py
"""Voucher verifier for the MPP session server.

Pure function: given a current channel snapshot and a signed voucher, decide
whether to accept (and what the new watermark would be), reject, or treat as an
idempotent replay. The caller persists any accepted delta through the channel
store, re-checking inside the atomic mutator.

The check sequence (order and operators) is normative and must be applied in
exactly this order::

    parse u64 -> sealed -> close pending -> idempotent replay (same
    cumulative AND same signature, signature re-verified) -> cumulative >
    watermark strictly -> cumulative <= deposit -> delta >= min_voucher_delta ->
    Ed25519 verify against the stored authorized_signer -> expiry.

Expiry follows the on-chain program's ``settle_and_seal`` rule
(``payment_channels`` ``helpers/voucher.rs``): the program rejects a voucher
only when ``expires_at != 0 && now >= expires_at``, so ``expires_at == 0`` means
**never-expires** and is always accepted. Off-chain acceptance mirrors that, with
one extra guard: because settlement is asynchronous (the server may co-sign and
broadcast the close up to the channel's forced-close grace period later), a
non-zero ``expires_at`` must also OUTLAST the settlement window — we reject when
``expires_at < now + settlement_window`` so a voucher cannot expire on-chain
after the server has already served the request but before the settle lands.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from enum import StrEnum

from solana_pay_kit.protocols.mpp.intents.session import SignedVoucher


class VoucherVerifyStatus(StrEnum):
    """The outcome class of a voucher verification."""

    #: The voucher advanced the channel watermark.
    ACCEPTED = "accepted"

    #: An already-accepted voucher was re-submitted (idempotent).
    REPLAYED = "replayed"

    #: The voucher was rejected; see ``VoucherVerifyResult.reason``.
    REJECTED = "rejected"


class VoucherRejectReason(StrEnum):
    """A stable string tag for voucher rejections so the caller can map to HTTP
    statuses / log levels without parsing free text. The tag values are part of
    the wire contract and must not change.
    """

    #: The delta is below the configured minimum.
    BELOW_MIN_DELTA = "below-min-delta"

    #: A close was already requested.
    CHANNEL_CLOSE_PENDING = "channel-close-pending"

    #: The channel is already sealed.
    CHANNEL_SEALED = "channel-sealed"

    #: The cumulative does not strictly exceed the watermark.
    CUMULATIVE_NOT_MONOTONIC = "cumulative-not-monotonic"

    #: The cumulative exceeds the deposit cap.
    EXCEEDS_DEPOSIT = "exceeds-deposit"

    #: The voucher expiry is not in the future.
    EXPIRED = "expired"

    #: The non-zero voucher expiry is in the future but does not outlast the
    #: settlement window, so it could expire on-chain before the asynchronous
    #: close settlement lands.
    EXPIRES_BEFORE_SETTLEMENT = "expires-before-settlement"

    #: The cumulative does not parse as a u64.
    INVALID_CUMULATIVE = "invalid-cumulative"

    #: The Ed25519 signature check failed.
    INVALID_SIGNATURE = "invalid-signature"


@dataclass
class ChannelState:
    """The persisted state of a single payment channel from the server's point
    of view, as read by the voucher verifier.

    The voucher verifier only reads a subset of the full channel state; this
    dataclass carries the fields it needs.
    """

    #: The on-chain channel address (base58).
    channel_id: str

    #: The public key authorized to sign vouchers for this session (base58).
    authorized_signer: str

    #: The total deposit / approved amount locked for this session (base units).
    deposit: int = 0

    #: The highest cumulative amount accepted by the server (the settled
    #: watermark).
    cumulative: int = 0

    #: True once the channel has been sealed on-chain.
    sealed: bool = False

    #: The signature of the highest accepted voucher (base58). Stored for
    #: idempotent replay detection.
    highest_voucher_signature: str | None = None

    #: The expiry timestamp from the highest accepted voucher.
    highest_voucher_expires_at: int | None = None

    #: The Unix timestamp (seconds) when cooperative close was requested. Once
    #: set, no further vouchers are accepted.
    close_requested_at: int | None = None


@dataclass
class VoucherVerifyResult:
    """The verdict of :func:`verify_voucher_for_channel`.

    ``status`` selects which fields are meaningful: ``new_cumulative`` for
    accepted and replayed; ``new_expires_at`` and ``new_signature`` for accepted
    only; ``reason`` and ``detail`` for rejected only.
    """

    #: The outcome class.
    status: VoucherVerifyStatus

    #: The watermark to persist (accepted) or the existing watermark (replayed).
    new_cumulative: int = 0

    #: The expiry of the now-highest voucher (accepted only).
    new_expires_at: int = 0

    #: The signature to persist as ``highest_voucher_signature`` (accepted only,
    #: base58).
    new_signature: str = ""

    #: The stable rejection tag (rejected only).
    reason: VoucherRejectReason | None = None

    #: A human-readable rejection detail. Safe to log; not stable.
    detail: str = ""


@dataclass
class VerifyVoucherArgs:
    """The inputs to :func:`verify_voucher_for_channel`."""

    #: The channel snapshot, typically read just before calling.
    state: ChannelState

    #: The voucher being submitted.
    signed: SignedVoucher

    #: The authoritative deposit cap. Passed in (rather than read off ``state``)
    #: because some callers carry an updated cap after a recent top-up that has
    #: not yet been written back into the store.
    deposit: int = 0

    #: The optional minimum delta from the previous cumulative. Zero disables
    #: the check.
    min_voucher_delta: int = 0

    #: The settlement window in seconds: the channel forced-close grace period
    #: the server must survive between accepting a voucher and landing the
    #: on-chain close settlement. A non-zero voucher ``expires_at`` must outlast
    #: ``now + settlement_window`` or it is rejected (see
    #: :attr:`VoucherRejectReason.EXPIRES_BEFORE_SETTLEMENT`). Zero disables the
    #: outlast check (only ``expires_at <= now`` rejects a non-zero expiry).
    settlement_window: int = 0

    #: Overrides the clock (Unix seconds) for deterministic tests. ``None``
    #: defaults to the wall clock.
    now_seconds: int | None = None


def verify_voucher_for_channel(args: VerifyVoucherArgs) -> VoucherVerifyResult:
    """Verify a voucher against a channel snapshot.

    Returns a verdict; the caller is responsible for persisting any accepted
    delta via the channel store. The verifier is pure: no store, network, or
    clock side effects (the clock is injectable).
    """
    state = args.state
    signed = args.signed

    # 1. Parse new cumulative from the payload.
    try:
        new_cumulative = _parse_u64(signed.data.cumulative)
    except ValueError:
        return _voucher_reject(
            VoucherRejectReason.INVALID_CUMULATIVE,
            f"invalid cumulative in voucher: {signed.data.cumulative}",
        )

    # 2. Channel must not be sealed.
    if state.sealed:
        return _voucher_reject(
            VoucherRejectReason.CHANNEL_SEALED,
            f"channel {state.channel_id} is already sealed",
        )

    # 3. Channel must not be in close-pending.
    if state.close_requested_at is not None:
        return _voucher_reject(
            VoucherRejectReason.CHANNEL_CLOSE_PENDING,
            f"channel {state.channel_id} close is pending; no further vouchers accepted",
        )

    # 4. Idempotent replay: same cumulative AND same signature. The signature is
    # re-verified so a replay of a forged voucher cannot slip through.
    if (
        new_cumulative == state.cumulative
        and state.highest_voucher_signature is not None
        and state.highest_voucher_signature == signed.signature
    ):
        err = _verify_voucher_signature_bytes(signed, state.authorized_signer)
        if err is not None:
            return _voucher_reject(VoucherRejectReason.INVALID_SIGNATURE, err)
        expiry_reject = _check_voucher_expiry(
            signed.data.expires_at, _voucher_now(args.now_seconds), args.settlement_window
        )
        if expiry_reject is not None:
            return expiry_reject
        return VoucherVerifyResult(status=VoucherVerifyStatus.REPLAYED, new_cumulative=new_cumulative)

    # 5. Must strictly exceed the watermark (non-replay case).
    if new_cumulative <= state.cumulative:
        return _voucher_reject(
            VoucherRejectReason.CUMULATIVE_NOT_MONOTONIC,
            f"voucher cumulative {new_cumulative} must exceed watermark {state.cumulative}",
        )

    # 6. Must not exceed the deposit.
    if new_cumulative > args.deposit:
        return _voucher_reject(
            VoucherRejectReason.EXCEEDS_DEPOSIT,
            f"voucher cumulative {new_cumulative} exceeds deposit {args.deposit}",
        )

    # 7. Min delta check.
    delta = new_cumulative - state.cumulative
    if args.min_voucher_delta > 0 and delta < args.min_voucher_delta:
        return _voucher_reject(
            VoucherRejectReason.BELOW_MIN_DELTA,
            f"voucher delta {delta} is below minimum {args.min_voucher_delta}",
        )

    # 8. Verify the Ed25519 signature over the 50-byte canonical payload.
    err = _verify_voucher_signature_bytes(signed, state.authorized_signer)
    if err is not None:
        return _voucher_reject(VoucherRejectReason.INVALID_SIGNATURE, err)

    # 9. Expiry. ``expires_at == 0`` is never-expires (matches the on-chain
    # settle); a non-zero expiry must be in the future AND outlast the settlement
    # window. The caller may override now_seconds for deterministic tests.
    expiry_reject = _check_voucher_expiry(
        signed.data.expires_at, _voucher_now(args.now_seconds), args.settlement_window
    )
    if expiry_reject is not None:
        return expiry_reject

    return VoucherVerifyResult(
        status=VoucherVerifyStatus.ACCEPTED,
        new_cumulative=new_cumulative,
        new_expires_at=signed.data.expires_at,
        new_signature=signed.signature,
    )


def _voucher_reject(reason: VoucherRejectReason, detail: str) -> VoucherVerifyResult:
    """Build a rejected verdict."""
    return VoucherVerifyResult(status=VoucherVerifyStatus.REJECTED, reason=reason, detail=detail)


def _check_voucher_expiry(expires_at: int, now: int, settlement_window: int) -> VoucherVerifyResult | None:
    """Apply the on-chain-aligned voucher expiry policy.

    Returns a rejected :class:`VoucherVerifyResult` when the voucher is expired
    or expires before the settlement window, or ``None`` when the expiry is
    acceptable.

    Policy (mirrors ``payment_channels`` ``helpers/voucher.rs``):

    - ``expires_at == 0`` is **never-expires** and is always accepted (the
      on-chain settle only rejects ``expires_at != 0 && now >= expires_at``).
    - A non-zero ``expires_at <= now`` is already expired -> ``EXPIRED``.
    - A non-zero ``expires_at`` that is in the future but does not outlast
      ``now + settlement_window`` could expire on-chain before the asynchronous
      close settlement lands -> ``EXPIRES_BEFORE_SETTLEMENT``. ``settlement_window``
      of 0 disables this guard.
    """
    if expires_at == 0:
        return None
    if expires_at <= now:
        return _voucher_reject(VoucherRejectReason.EXPIRED, "voucher has expired")
    if settlement_window > 0 and expires_at < now + settlement_window:
        return _voucher_reject(
            VoucherRejectReason.EXPIRES_BEFORE_SETTLEMENT,
            f"voucher expires_at {expires_at} does not outlast the settlement window "
            f"(now {now} + {settlement_window}s = {now + settlement_window})",
        )
    return None


def _voucher_now(override: int | None) -> int:
    """Return the override when set, otherwise the wall clock in Unix seconds."""
    if override is not None:
        return override
    return int(time.time())


_U64_MAX = (1 << 64) - 1


def _parse_u64(raw: str) -> int:
    """Parse a canonical unsigned base-10 ``u64``.

    Rejects empty, signed, fractional, non-ASCII-digit, or out-of-range values.
    """
    if not (raw.isascii() and raw.isdigit()):
        raise ValueError(f"invalid cumulative {raw!r}")
    value = int(raw, 10)
    if value > _U64_MAX:
        raise ValueError(f"cumulative {raw!r} exceeds u64 range")
    return value


def _verify_voucher_signature_bytes(signed: SignedVoucher, authorized_signer: str) -> str | None:
    """Check the voucher's Ed25519 signature over the canonical 50-byte voucher
    payload against the authorized signer (both base58). The expiry check is not
    included; callers order it explicitly.

    Returns ``None`` on success or a human-readable error string on failure.
    """
    from solders.pubkey import Pubkey  # type: ignore[import-untyped]
    from solders.signature import Signature  # type: ignore[import-untyped]

    try:
        message = signed.data.message_bytes()
    except ValueError as exc:
        return str(exc)
    try:
        signature = Signature.from_string(signed.signature)
    except (ValueError, TypeError) as exc:
        return f"invalid signature encoding: {exc}"
    try:
        pubkey = Pubkey.from_string(authorized_signer)
    except (ValueError, TypeError) as exc:
        return f"invalid authorized signer: {exc}"
    if not signature.verify(pubkey, message):
        return "voucher signature verification failed"
    return None


def verify_session_voucher(signed: SignedVoucher, authorized_signer: str, settlement_window: int = 0) -> str | None:
    """Check expiry (including the settlement-window margin), then the signature.

    Used by the commit and close paths; the voucher handler orders the two
    checks itself. Returns ``None`` on success or a human-readable error string
    on failure.

    Expiry mirrors the on-chain settle rule and the voucher path's
    :func:`_check_voucher_expiry`: ``expires_at == 0`` is never-expires; a
    non-zero ``expires_at <= now`` has expired; and when ``settlement_window``
    is positive, a voucher that does not outlast ``now + settlement_window`` is
    rejected so a committed or closing voucher cannot expire before the async
    settle transaction lands on-chain.
    """
    rejection = _check_voucher_expiry(signed.data.expires_at, int(time.time()), settlement_window)
    if rejection is not None:
        return rejection.detail
    return _verify_voucher_signature_bytes(signed, authorized_signer)
